File: DriftGame/uracer-desktop/src/com/bitfire/uracer/game/logic/types/helpers/TimeModulator.java
package com.bitfire.uracer.game.logic.types.helpers;

import com.bitfire.uracer.game.tween.SysTweener;
import com.bitfire.uracer.utils.AMath;
import com.bitfire.uracer.utils.BoxedFloat;
import com.bitfire.uracer.utils.BoxedFloatAccessor;

import aurelienribon.tweenengine.Timeline;
import aurelienribon.tweenengine.Tween;
import aurelienribon.tweenengine.equations.Quad;

public final class TimeModulator {

    public static final float MinTime = 0.25f;
    public static final float MaxTime = 1.0f;

    private final BoxedFloat timeMultiplier;
    private Timeline timeSeq;

    public TimeModulator() {
        timeMultiplier = new BoxedFloat(1);
        timeSeq = Timeline.createSequence();
    }

    // returns the modulate time value
    public float getTime() {
        return AMath.clamp(timeMultiplier.value, MinTime, MaxTime);
    }

    public void reset() {
        timeMultiplier.value = MaxTime;
    }

    public void toDilatedTime() {
        modulateTo(MinTime);
    }

    public void toNormalTime() {
        modulateTo(MaxTime);
    }

    private void modulateTo(float to) {
        SysTweener.stop(timeMultiplier);
        timeSeq = Timeline.createSequence();
        timeSeq.push(Tween.to(timeMultiplier, BoxedFloatAccessor.VALUE, (float) 1000).target(to).ease(Quad.OUT));
        SysTweener.start(timeSeq);
    }
}

File: DriftGame/uracer-desktop/src/com/bitfire/uracer/game/logic/types/helpers/TimeModulator.kt
package com.bitfire.uracer.game.logic.types.helpers

import aurelienribon.tweenengine.Timeline
import aurelienribon.tweenengine.Tween
import aurelienribon.tweenengine.equations.Quad.Companion.OUT
import com.bitfire.uracer.game.tween.SysTweener.start
import com.bitfire.uracer.game.tween.SysTweener.stop
import com.bitfire.uracer.utils.AlgebraMath.clamp
import com.bitfire.uracer.utils.BoxedFloat
import com.bitfire.uracer.utils.BoxedFloatAccessor

class TimeModulator {

    private val timeMultiplier = BoxedFloat(1f)
    private var timeSeq = Timeline.createSequence()

    /**
     * returns the modulate time value.
     */
    val time = clamp(timeMultiplier.value, MIN_TIME, MAX_TIME)

    fun reset() {
        timeMultiplier.value = MAX_TIME
    }

    fun toDilatedTime() {
        modulateTo(MIN_TIME)
    }

    fun toNormalTime() {
        modulateTo(MAX_TIME)
    }

    private fun modulateTo(to: Float) {
        stop(timeMultiplier)
        timeSeq = Timeline.createSequence()
        timeSeq.push(Tween.to(timeMultiplier, BoxedFloatAccessor.VALUE, 1000f).target(to).ease(OUT))
        start(timeSeq)
    }

    companion object {
        const val MIN_TIME = 0.25F
        const val MAX_TIME = 1.0F
    }
}
